src/equations/serre_green_naghdi_flat_1d.jl

```diff
@@ -0,0 +1,456 @@
+                   0.5 * h * v * v_x
+                   +
+                   p_x)
+    end
+
+    @trixi_timeit timer() "assembling elliptic operator" begin
+        # The code below is equivalent to
+        #   dv .= (Diagonal(h) - Dmat * Diagonal(1/3 .* h.^3) * Dmat) \ tmp
+        # but faster since the symbolic factorization is reused.
+        # Floating point errors accumulate a bit and the system matrix
+        # is not necessarily perfectly symmetric but only up to round-off
+        # errors. We wrap it here to avoid issues with the factorization.
+        @. M_h = h
+        scale_by_mass_matrix!(M_h, D)
+        inv3 = 1 / 3
+        @. M_h3_3 = inv3 * h^3
+        scale_by_mass_matrix!(M_h3_3, D)
+        system_matrix = Symmetric(Diagonal(M_h)
+                                  +
+                                  Dmat' * Diagonal(M_h3_3) * Dmat)
+    end
+
+    @trixi_timeit timer() "solving elliptic system" begin
+        if issparse(system_matrix)
+            (; factorization) = cache
+            cholesky!(factorization, system_matrix; check = false)
+            if issuccess(factorization)
+                scale_by_mass_matrix!(tmp, D)
+                dv .= factorization \ tmp
+            else
+                # The factorization may fail if the time step is too large
+                # and h becomes negative.
+                fill!(dv, NaN)
+            end
+        else
+            factorization = cholesky!(system_matrix)
+            scale_by_mass_matrix!(tmp, D)
+            ldiv!(dv, factorization, tmp)
+        end
+    end
+
+    return nothing
+end
+
+function rhs_sgn_flat_upwind!(dq, q, equations, source_terms, cache)
+    # Unpack physical parameters and SBP operator `D` as well as the
+    # SBP upwind operator in sparse matrix form `Dmat_minus`
+    g = equations.gravity
+    (; Dmat_minus) = cache
+    D_upwind = cache.D
+    D = D_upwind.central
+
+    # `q` and `dq` are `ArrayPartition`s. They collect the individual
+    # arrays for the water height `h` and the velocity `v`.
+    h, v = q.x
+    dh, dv = dq.x
+
+    @trixi_timeit timer() "hyperbolic terms" begin
+        # Compute all derivatives required below
+        (; h_x, v_x, v_x_upwind, h2_x, hv_x, v2_x,
+        h2_v_vx_x, h_vx_x, p_x, tmp,
+        M_h, M_h3_3) = cache
+
+        mul!(h_x, D, h)
+        mul!(v_x, D, v)
+        mul!(v_x_upwind, D_upwind.minus, v)
+        @. tmp = h^2
+        mul!(h2_x, D, tmp)
+        @. tmp = h * v
+        mul!(hv_x, D, tmp)
+        @. tmp = v^2
+        mul!(v2_x, D, tmp)
+
+        @. tmp = h^2 * v * v_x
+        mul!(h2_v_vx_x, D, tmp)
+        @. tmp = h * v_x
+        mul!(h_vx_x, D, tmp)
+        # p_+
+        @. tmp = 0.5 * h^2 * (h * v_x + h_x * v) * v_x_upwind
+        mul!(p_x, D_upwind.plus, tmp)
+        # p_0
+        minv6 = -1 / 6
+        @. tmp = minv6 * (h * h2_v_vx_x
+                          +
+                          h^2 * v * h_vx_x)
+        mul!(p_x, D, tmp, 1.0, 1.0)
```

Comment:
```suggestion
        mul!(p_x, D, tmp)
```
Why do you use the 5-argument `mul!` when you pass ones anyway?

Reply:
Since the default values are `1, 0` and not `1, 1`:
```julia
julia> C = [1.0;;]; A = [2.0;;]; B = [3.0;;]; mul!(C, A, B)
1×1 Matrix{Float64}:
 6.0

julia> C = [1.0;;]; A = [2.0;;]; B = [3.0;;]; mul!(C, A, B, 1, 1)
1×1 Matrix{Float64}:
 7.0
```

Reply:
Ah, sure. My bad. Sorry.